### PR TITLE
[ROCm] [ROCm2.6+] Enable jit fusion on ROCm

### PR DESCRIPTION
**Pull request by iotamudelta**
As of ROCm 2.6, we support hiprtc - the HIP runtime compilation API. Enable the jit fusion feature depending on the existence of such an API. This entails
* new hipification rules for API_RTC
* add hiprtc APIs to the shim loader
* update cmake infrastructure to find the hiprtc library (it is part of the HIP package)
* enabling of unit tests in the jit_fuser test set
* special casing in resource strings for HIP - the typedefs CUDA requires would be redundant
* for now disable the occupancy calculation we do not support yet and hard-code

Thanks to @t-vi for working with me on getting this integration done!

### CODE REVIEW

File: torch/csrc/jit/fuser/cuda/fused_kernel.cpp [hunk 1]
```diff
@@ -108,13 +108,21 @@ FusedKernelCUDA::FusedKernelCUDA(
...
+#else
   const std::string compute = "--gpu-architecture=compute_" +
       std::to_string(major) + std::to_string(minor);
   const std::vector<const char*> args = {
       "--std=c++11", compute.c_str(), "-default-device"};
+#endif
   const auto result =
       nvrtc().nvrtcCompileProgram(program, args.size(), args.data());
+#ifdef __HIP_PLATFORM_HCC__
+  if (result != HIPRTC_SUCCESS) {
```
Comment (bddppq) [inline torch/csrc/jit/fuser/cuda/fused_kernel.cpp]: Could this be hipified?

**Review by bddppq — COMMENTED** [review]
Wow awesome!

**Comment by bddppq** [comment]
Triggered CI jobs with ROCm 2.6 docker images:
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger/39291/
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-devtoolset7-rocmrpm-centos7.5-trigger/17246/
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger-test/32007/
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-devtoolset7-rocmrpm-centos7.5-trigger-test/32336/

File: torch/csrc/jit/fuser/cuda/fused_kernel.cpp (hunk 1 above)
Comment (iotamudelta) [reply torch/csrc/jit/fuser/cuda/fused_kernel.cpp]: Only if I can switch the CUDA path to also do the catch-all for `!= NVRTC_SUCCESS`.

**Comment by iotamudelta** [comment]
I believe the 2.6 images are rocThrust images - so this won't work w/o merging the rocThrust PR in. :-)

**Comment by bddppq** [comment]
ehh ... ok sorry then we need to put this on hold until we upgrade to ROCm 2.6

File: torch/csrc/jit/fuser/cuda/fused_kernel.cpp (hunk 1 above)
Comment (bddppq) [reply torch/csrc/jit/fuser/cuda/fused_kernel.cpp]: Yes I think changing to do catch-all is reasonable.

**Comment by iotamudelta** [comment]
I figured as much. I just wanted to make you guys aware that the integration is done and wanted to ask for review so that it's ready to test and land once the CI is on 2.6 or higher.

File: torch/csrc/jit/fuser/cuda/resource_strings.h [hunk 2]
```diff
@@ -13,6 +13,25 @@ tensor as input. Correct code for this case is generated, however, nvrtc does
 not know how to handle int*_t integer types, so typedefs help it handle those
 cases*/
 
+#ifdef __HIP_PLATFORM_HCC__
+static auto type_declarations_template = CodeTemplate(R"(
+#include <hip/hip_runtime.h>
+${HalfHeader}
+${RandHeader}
+
```
Comment (t-vi) [inline torch/csrc/jit/fuser/cuda/resource_strings.h]: I think you want the `POS_INFINITY` and `NEG_INFINITY` constant definitions (maybe just to `INFINITY` and `-INFINITY`) to pass the `TestFuser.test_clamp` test. (Of course, one might see if we strictly need to emit `(a < INFINITY ? a : INFINITY)`, but that's a different change.)

**Review by ezyang — APPROVED** [review]
Looks legit.

File: torch/csrc/jit/fuser/cuda/resource_strings.h (hunk 2 above)
Comment (iotamudelta) [reply torch/csrc/jit/fuser/cuda/resource_strings.h]: Spot on, thanks!

**Comment by ezyang** [comment]
Needs rebasing, unfortunately :(

**Comment by iotamudelta** [comment]
looks like merging did the trick. thanks for pinging!

**Comment by ezyang** [comment]
More conflicts :(

**Comment by iotamudelta** [comment]
@bddppq this should be ready now?

**Review by bddppq — APPROVED** [review]
LGTM

**Comment by bddppq** [comment]
cc @xw285cornell

**Comment by iotamudelta** [comment]
@bddppq looks like the C2 testing is running into the CI time limit?

**Comment by iotamudelta** [comment]
@ezyang can we bump the CI time limit?

**Comment by bddppq** [comment]
Bumped to 120 minutes

@pytorchbot retest this please

**Comment by iotamudelta** [comment]
@bddppq now that looks better! :-) thanks!

**Comment by bddppq** [comment]
@iotamudelta Could you rebase to re-trigger the failed cuda CI jobs? pytorchbot command doesn't seem to be able to make it correctly rerun. The reason I want to see them green is in the first run they showed memory leaking errors in one of the jit test (test_lerp_cuda), while I couldn't see any changes in this PR that can affect cuda build.

**Comment by iotamudelta** [comment]
@bddppq remerged. The only change to the CUDA path is https://github.com/pytorch/pytorch/pull/22872/commits/2cf85dedf4b99f2660f813e5aa0e419a7f502432

**Comment by bddppq** [comment]
@iotamudelta Thanks. The tests are passing now, I'm going to land this soon.

**Pull request closed without merging**